### PR TITLE
Feat kind support for AF_XDP device plugin

### PR DESCRIPTION
**Pull request by maryamtahhan**
depends on https://github.com/intel/afxdp-plugins-for-kubernetes/pull/57 in order to build

- [x] Cleanup any dead code
- [x] Update the documentation
- [X] Re-test CNDP with a physical cluster - not needed as we removed the cluster type flag

This PR enables support for deployment of the AF_XDP device plugin in kind. It's been tested with CNDP and [updated CNDP](https://github.com/CloudNativeDataPlane/cndp/pull/281) to be able to work with the current implementation of the DP.

**Note:** one limitation for now is that the Pod requires CAP_BPF in the Kind setup. The limitation was due to the ` kernel.unprivileged_bpf_disabled` kernel flag being set by default in the kind node. Once this was disabled everything worked as expected.

The main value prop here is to enable the spin up of a test environment in a matter of minutes as well as being a building block towards integration with BPFd

### CODE REVIEW

File: examples/kind-pod-spec.yaml [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+  - name: afxdp
+    image: docker-image:latest                 # Specify your docker image here, along with PullPolicy and command
+    imagePullPolicy: IfNotPresent
+    command: ["tail", "-f", "/dev/null"]
+    securityContext:
+      capabilities:
+        add:
+          - NET_RAW
+          - IPC_LOCK
+          - BPF
```
Comment (maryamtahhan) [inline examples/kind-pod-spec.yaml:20]: add a comment that this is needed for a kind pod only!!

File: images/amd64.dockerfile [hunk 2]
```diff
@@ -2,9 +2,9 @@
 # Licensed under the Apache License, Version 2.0 (the "License");
```
Comment (maryamtahhan) [inline images/amd64.dockerfile:2]: update copyright tag

File: internal/bpf/xdp-pass/xdp_pass.c [hunk 3]
```diff
@@ -0,0 +1,10 @@
+#include <linux/bpf.h>
```
Comment (maryamtahhan) [inline internal/bpf/xdp-pass/xdp_pass.c:1]: copyright tag

File: internal/networking/kind.go [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package networking
+
+import (
+	"strconv"
+//	"net"
```
Comment (maryamtahhan) [inline internal/networking/kind.go:20]: remove

File: internal/networking/networking.go [hunk 5]
```diff
@@ -415,6 +418,17 @@ func readDeviceMap(filepath string) (map[string]*DeviceDetails, error) {
 	return deviceMap, nil
 }
 
+/*
```
Comment (maryamtahhan) [inline internal/networking/networking.go:421]: delete this

**Comment by maryamtahhan** [comment]
Btw the changes to the makefile allow you to type make help and you will see the following:

``` bash
# make help

Usage:
  make <target>
  help             Display this help.

General Build - assumes K8s environment is already setup
  docker           Build docker image
  podman           Build podman image
  undeploy         Undeploy the Deamonset
  deploy           Deploy the Deamonset and CNI

General setup
  setup-flannel    Setup flannel
  setup-multus     Setup multus

Kind Deployment - sets up a kind cluster and deploys the plugin and CNI
  del-kind         Remove a kind cluster callled af-xdp-deployment
  setup-kind       Setup a kind cluster called af-xdp-deployment
  label-kind-nodes  label the kind worker nodes with cndp="true"
  kind-deploy      Deploy the Deamonset and CNI in Kind
  run-on-kind      Setup a kind cluster and deploy the device plugin
  ```

File: README.md [hunk 6]
```diff
@@ -1,129 +1,155 @@
...
-	sudo apt-get update
-	sudo apt-get install trivy
-	```
+  - A comprehensive and versatile security scanner.
+  - [Install on Ubuntu](https://aquasecurity.github.io/trivy/v0.38/getting-started/installation/):
+
+    ```bash
+    sudo apt-get install wget apt-transport-https gnupg lsb-release
+    wget -qO - https://aquasecurity.github.io/trivy-repo/deb/public.key | gpg --dearmor | sudo tee /usr/share/keyrings/trivy.gpg > /dev/null
+    echo "deb [signed-by=/usr/share/keyrings/trivy.gpg] https://aquasecurity.github.io/trivy-repo/deb $(lsb_release -sc) main" |                     sudo tee -a /etc/apt/sources.list.d/trivy.list
```
Comment (maryamtahhan) [inline README.md:102]: double check this

**Comment by maryamtahhan** [comment]
Tested with CNDP

```
root@cndp-0-0:~# cndpfwd -c config.jsonc -L debug drop
(_lgroup                 :  69) DEBUG: CPUs configured 4 available 4
(_lgroup                 :  69) DEBUG: CPUs configured 4 available 4
(_lgroup                 :  69) DEBUG: CPUs configured 4 available 4
(_lgroup                 :  69) DEBUG: CPUs configured 4 available 4
(_lgroup                 :  69) DEBUG: CPUs configured 4 available 4
(_lgroup                 :  69) DEBUG: CPUs configured 4 available 4
(mmap_alloc              : 203) WARNING: unable to allocate 2MB hugepages, trying 4KB pages
(cne_ring_init           : 221) DEBUG: ring is not cache aligned r=0x7f9af0f5c010 aligned=0x7f9af0f5c040
(xskdev_recv_xsk_fd      : 608) DEBUG: Received /host_ok msg
(xskdev_recv_xsk_fd      : 616) DEBUG: Sent /xsk_map_fd,veth10 msg
(xskdev_recv_xsk_fd      : 625) DEBUG: waiting for the xsk_map_fd the num of tries is 1
(xskdev_recv_xsk_fd      : 628) DEBUG: the xsk_map_fd is 4
(xskdev_socket_create    : 762) INFO: xi->xsk_map_fd = 4
libbpf: can't get next link: Operation not permitted
(xskdev_socket_create    : 910) DEBUG: Program ID 0, if_index 8, if_name 'veth10'

*** CNDPFWD Forward Application, API: XSKDEV, Mode: Drop, Burst Size: 256
   Initial Thread ID    36 on lcore 0
   Forwarding Thread ID 42 on lcore 0
  ```

**Comment by maryamtahhan** [comment]
I'll have a look through the CI and try to fix

File: README.md [hunk 7]
```diff
@@ -423,10 +467,38 @@ The example below shows a config including log settings.
 }
 ```
 
+### Cluster Types
+
+The cluster type flag is used to indicate if this is a physical cluster or a Kind cluster.
+
+```yaml
+{
+       "clusterType":"physical",
```
Comment (maryamtahhan) [inline README.md:476]: There is an option to just make this key something more kind specific to reduce the impact of having to change this configuration for any existing clusters....

Comment (maryamtahhan) [reply README.md]: removed it

File: README.md [hunk 8]
```diff
@@ -1,146 +1,184 @@
...
+This will deploy a kind cluster with a control plane and two worker nodes. It will build and run the DP daemonset, and install the CNI on all nodes.
+
+### Running Pods
+
+- Create a network attachment definition file. This is the config for the CNI plugin.
+  - An example file can be found under [examples/network-attachment-definition.yaml](./examples/network-attachment-definition.yaml)
+  - Change the config if necessary. See comments in the example file.
+  - `kubectl create -f network-attachment-definition.yaml`
+- Create a pod spec:
+  - An example pod spec can be found under [examples/pod-spec.yaml](./examples/pod-spec.yaml)
```
Comment (maryamtahhan) [inline README.md:166]:
```suggestion
  - An example pod spec can be found under [examples/kind-pod-spec.yaml](./examples/kind-pod-spec.yaml)
```

File: README.md [hunk 9]
```diff
@@ -1,146 +1,184 @@
...
+  - Configure the pod spec to use a suitable Docker image and to reference the network attachment definition as well as the resource type from the Device Plugin. See comments in the example file.
+  - `kubectl create -f pod-spec.yaml`
```
Comment (maryamtahhan) [inline README.md:168]:
```suggestion
  - `kubectl create -f kind-pod-spec.yaml`
```

File: README.md [hunk 10]
```diff
@@ -341,67 +389,73 @@ However, there are also three node-specific configs included:
...
 
 #### UdsServerDisable
+
 UdsServerDisable is a Boolean configuration. If set to true, devices in this pool will not have the BPF app loaded onto the netdev. This means no UDS server is spun up when a device is allocated to a pod. By default, this is set to false.
 
 #### UdsTimeout
-UdsTimeout is an integer configuration. This value sets the amount of time, in seconds, that the UDS server will wait while there is no activity on the UDS. When this timeout limit is reached, the UDS server terminates and the UDS is deleted from the filesystem. This can be a useful setting, for example, in scenarios where large batches of pods are created together. Large batches of pods tend to take some time to spin up, so it might be beneficial to have the UDS server sit waiting a little longer for the pod to start. The maximum allowed value is 300 seconds (5 min). The minimum and default value is 30 seconds.
 
-#### RequiresUnprivilegedBpf
```
Comment (maryamtahhan) [inline README.md:359]: Add this back

File: README.md [hunk 11]
```diff
@@ -1,146 +1,184 @@
...
-	- `kubectl create -f pod-spec.yaml`
+  - An example pod spec can be found under [examples/pod-spec.yaml](./examples/pod-spec.yaml)
+  - Configure the pod spec to use a suitable Docker image and to reference the network attachment definition as well as the resource type from the Device Plugin. See comments in the example file.
+  - `kubectl create -f pod-spec.yaml`
 
 ## Prerequisites
+
 ### Running the Plugins
+
 The following prerequisites are required to run the plugins:
```
Comment (maryamtahhan) [inline README.md:33]: What about kind?

**Comment by maryamtahhan** [comment]
Hey @garyloug 

This change has been rebased on the latest main and is ready for review. It has been tested with CNDP. 

In the meantime I will move on to complete the BPF map pinning changes.

Thanks

File: internal/deviceplugin/configFile.go [hunk 12]
```diff
@@ -99,6 +101,8 @@ type configFile struct {
 	Pools    []*configFile_Pool `json:"Pools"`
 	LogFile  string             `json:"LogFile"`
 	LogLevel string             `json:"LogLevel"`
+	// ClusterType string             `json:"clusterType"`
```
Comment (maryamtahhan) [inline internal/deviceplugin/configFile.go:104]: cleanup

**Comment by maryamtahhan** [comment]
I will cleanup the suggestions that were put forward by the linter bot...

I think that the kind configuration/support as it's used for testing only - should be limited to a single pool. I will also update the code to make this explicit...

**Comment by garyloug** [comment]
hey @maryamtahhan - is this one still relevant or superseded by "v2" #59 ?

**Comment by maryamtahhan** [comment]
This is just the kind support...
https://github.com/intel/afxdp-plugins-for-kubernetes/pull/59 builds on top of it to add bpf map pinning... up to you if you prefer to review all part of 1 change or to do things staged...

if you review the kind support here alone - the bpf map pinning changes PR will become significantly smaller

**Review by garyloug — DISMISSED** [review]

**Review by garyloug — APPROVED** [review]

**Pull request merged**